tests/unit/s2n_cert_validation_callback_test.c

```diff
@@ -0,0 +1,414 @@
+    } else {
+        EXPECT_SUCCESS(s2n_cert_validation_reject(info));
+    }
+
+    return ret;
+}
+
+static int test_cert_validation_callback_self_talk(struct s2n_connection *conn,
+        struct s2n_cert_validation_info *info, void *ctx)
+{
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key *peer_cert_chain = s2n_cert_chain_and_key_new(),
+            s2n_cert_chain_and_key_ptr_free);
+    EXPECT_NOT_NULL(peer_cert_chain);
+
+    /* Ensure that the peer's certificate chain can be retrieved at the time the callback is invoked */
+    EXPECT_SUCCESS(s2n_connection_get_peer_cert_chain(conn, peer_cert_chain));
+    uint32_t peer_cert_chain_len = 0;
+    EXPECT_SUCCESS(s2n_cert_chain_get_length(peer_cert_chain, &peer_cert_chain_len));
+    EXPECT_TRUE(peer_cert_chain_len > 0);
+
+    return test_cert_validation_callback(conn, info, ctx);
+}
+
+static int test_cert_validation_callback_self_talk_server(struct s2n_connection *conn,
+        struct s2n_cert_validation_info *info, void *ctx)
+{
+    /* Ensure that the callback was invoked on the server connection */
+    EXPECT_EQUAL(conn->mode, S2N_SERVER);
+
+    /* Ensure that the client's certificate chain can be retrieved at the time the callback was invoked */
+    uint8_t *der_cert_chain = 0;
+    uint32_t cert_chain_len;
+    EXPECT_SUCCESS(s2n_connection_get_client_cert_chain(conn, &der_cert_chain, &cert_chain_len));
+    EXPECT_TRUE(cert_chain_len > 0);
+
+    return test_cert_validation_callback_self_talk(conn, info, ctx);
+}
+
+static int test_cert_validation_callback_self_talk_ocsp(struct s2n_connection *conn,
+        struct s2n_cert_validation_info *info, void *ctx)
+{
+    /* Ensure that the OCSP response was received prior to invoking the callback */
+    uint32_t ocsp_response_length = 0;
+    const uint8_t *ocsp_response = s2n_connection_get_ocsp_response(conn, &ocsp_response_length);
+    EXPECT_NOT_NULL(ocsp_response);
+    EXPECT_TRUE(ocsp_response_length > 0);
+
+    return test_cert_validation_callback_self_talk(conn, info, ctx);
+}
+
+int main(int argc, char *argv[])
+{
+    BEGIN_TEST();
+
+    /* Accept/reject tests */
+    {
+        /* Safety */
+        EXPECT_FAILURE_WITH_ERRNO(s2n_cert_validation_accept(NULL), S2N_ERR_NULL);
+        EXPECT_FAILURE_WITH_ERRNO(s2n_cert_validation_reject(NULL), S2N_ERR_NULL);
+
+        /* Accept sets the proper state */
+        {
+            struct s2n_cert_validation_info info = { 0 };
+
+            EXPECT_SUCCESS(s2n_cert_validation_accept(&info));
+
+            EXPECT_EQUAL(info.finished, true);
+            EXPECT_EQUAL(info.accepted, true);
+        }
+
+        /* Reject sets the proper state */
+        {
+            struct s2n_cert_validation_info info = { 0 };
+
+            EXPECT_SUCCESS(s2n_cert_validation_reject(&info));
+
+            EXPECT_EQUAL(info.finished, true);
+            EXPECT_EQUAL(info.accepted, false);
+        }
+
+        /* Calls to accept/reject fail if accept has already been called */
```

Comment:
Why is calling this API twice an error case?

Reply:
It seemed unnecessary to call the function twice, and also not entirely clear what the expected behavior should be if you call accept and then reject. Assuming we implemented it to override whatever the previous call was, this could be weird in an async context where the first call is made and then `s2n_negotiate()` is re-entered somehow before the next call is made, leading to undesired behavior. So it seemed safer to just error. Does it seems clear/useful to allow calling the API twice? If so I can remove the check.

Reply:
Nah, that's a good reason.